### PR TITLE
Fix abstract type federation. PR includes dnerdy's (Mark Sandstorm) work

### PR DESCRIPTION
**Pull request by Aenimus**
Added missing __typename to a variety of places including interfaces and unions.
 
 Original PR: https://github.com/wundergraph/graphql-go-tools/pull/334/files
 
 Needs more tests to prove the changes are robust.

### CODE REVIEW

File: pkg/engine/datasource/graphql_datasource/graphql_datasource.go [hunk 1]
```diff
@@ -686,8 +692,9 @@ func (p *Planner) fieldDefinition(fieldName, typeName string) *ast.FieldDefiniti
 	return &p.visitor.Definition.FieldDefinitions[definition]
 }
 
-func (p *Planner) addOneTypeInlineFragment() {
+func (p *Planner) addOnTypeInlineFragment() {
 	selectionSet := p.upstreamOperation.AddSelectionSet()
+	p.addTypenameToSelectionSet(p.nodes[len(p.nodes)-1].Ref)
```
Comment (Aenimus) [inline pkg/engine/datasource/graphql_datasource/graphql_datasource.go:738]: This is where __typename is added to the fragment.

File: pkg/graphql/execution_engine_v2_test.go [hunk 2]
```diff
@@ -1163,7 +1163,7 @@ func TestExecutionEngineV2_Execute(t *testing.T) {
 						HTTPClient: testNetHttpClient(t, roundTripperTestCase{
 							expectedHost:     "example.com",
 							expectedPath:     "/",
-							expectedBody:     `{"query":"{codeType {code __typename ... on Country {name}}}"}`,
+							expectedBody:     `{"query":"{codeType {__typename code ... on Country {name}}}"}`,
```
Comment (Aenimus) [inline pkg/graphql/execution_engine_v2_test.go:1320]: The __typename swapped places in this instance. Is that a problem?

Comment (jensneuse) [reply pkg/graphql/execution_engine_v2_test.go:1320]: No

File: pkg/engine/plan/plan.go [hunk 3]
```diff
@@ -1390,9 +1393,19 @@ type configurationVisitor struct {
 	currentBufferId       int
 	fieldBuffers          map[int]int
 
+	parentTypeNodes []ast.Node
+
```
Comment (Aenimus) [inline pkg/engine/plan/plan.go:1396]: What is the purpose of this spacing? @jensneuse

File: examples/federation/reviews/graph/schema.graphqls [hunk 4]
```diff
@@ -6,7 +6,6 @@ type Review {
 
 extend type User @key(fields: "id") {
     id: ID! @external
-    username: String! @external
```
Comment (Aenimus) [inline examples/federation/reviews/graph/schema.graphqls]: Username is unnecessary to include on the entity stub.

**Review by jensneuse — APPROVED** [review]

**Review by spetrunin — APPROVED** [review]

**Pull request merged**